manifold/src/impl.cpp

```diff
@@ -525,47 +528,74 @@ void Manifold::Impl::Update() {
   collider_.UpdateBoxes(faceBox);
 }
 
-void Manifold::Impl::ApplyTransform() const {
-  // This const_cast is here because these operations cancel out, leaving the
-  // state conceptually unchanged. This enables lazy transformation evaluation.
-  const_cast<Impl*>(this)->ApplyTransform();
-}
-
 /**
- * Bake the manifold's transform into its vertices. This function allows lazy
- * evaluation, which is important because often several transforms are applied
- * between operations.
+ * Bake the manifold's transform into its vertices.
  */
-void Manifold::Impl::ApplyTransform() {
-  if (transform_ == glm::mat4x3(1.0f)) return;
-  auto policy = autoPolicy(vertPos_.size());
-  for_each(policy, vertPos_.begin(), vertPos_.end(),
-                   Transform4x3({transform_}));
+void Manifold::Impl::ApplyTransform(const glm::mat4x3 &transform_) {
+  if (transform_ == glm::mat4x3(1.0f))
+    return;
+  auto policy = autoPolicy(NumVert());
+  transform(policy, vertPos_.begin(), vertPos_.end(), vertPos_.begin(),
+            Transform4x3({transform_}));
 
   glm::mat3 normalTransform =
       glm::inverse(glm::transpose(glm::mat3(transform_)));
-  for_each(policy, faceNormal_.begin(), faceNormal_.end(),
-                   TransformNormals({normalTransform}));
-  for_each(policy, vertNormal_.begin(), vertNormal_.end(),
-                   TransformNormals({normalTransform}));
+  transform(policy, faceNormal_.begin(), faceNormal_.end(), faceNormal_.begin(),
+            TransformNormals({normalTransform}));
+  transform(policy, vertNormal_.begin(), vertNormal_.end(), vertNormal_.begin(),
+            TransformNormals({normalTransform}));
   // This optimization does a cheap collider update if the transform is
   // axis-aligned.
-  if (!collider_.Transform(transform_)) Update();
+  if (!collider_.Transform(transform_))
+    Update();
 
   const float oldScale = bBox_.Scale();
-  transform_ = glm::mat4x3(1.0f);
   CalculateBBox();
 
   const float newScale = bBox_.Scale();
-  precision_ *= glm::max(1.0f, newScale / oldScale) *
-                glm::max(glm::length(transform_[0]),
-                         glm::max(glm::length(transform_[1]),
-                                  glm::length(transform_[2])));
```

Comment:
I think this is somehow causing the Precision test to fail. That's a real failure; it's testing that objects below the floating-point precision limit get collapsed to nothing. This could easily be causing the other failures on the big tests too.

Reply:
Yes, I think I tried using the transform size before resetting it but got some errors, I will try to change this and see how it goes.

Reply:
I found an error in my patch: forgot to use `result.precision_` in `result.SetPrecision`.

Adding these lines back triggered a few more errors (`Boolean.Precision2`, `Manifold.Precision`, `Samples.TetPuzzle`).
```diff
 Manifold::Impl Manifold::Impl::Transform(const glm::mat4x3 &transform_) const {
   if (transform_ == glm::mat4x3(1.0f)) return *this;
   auto policy = autoPolicy(NumVert());
@@ -589,13 +558,15 @@ Manifold::Impl Manifold::Impl::Transform(const glm::mat4x3 &transform_) const {
   result.CalculateBBox();
 
   const float newScale = result.bBox_.Scale();
-  result.precision_ *= glm::max(1.0f, newScale / oldScale);
+  result.precision_ *= glm::max(1.0f, newScale / oldScale) *
+                       glm::max(glm::length(transform_[0]),
+                                glm::max(glm::length(transform_[1]),
+                                         glm::length(transform_[2])));
 
   // Maximum of inherited precision loss and translational precision loss.
-  result.SetPrecision(precision_);
+  result.SetPrecision(result.precision_);
   return result;
 }
```

Reply:
Not sure how clear my code is here, but the idea is that for geometric intersections, floating point precision doesn't make a lot of sense. Fixed point would be better, but that's not how modern chips are made, so I just keep track of a fixed precision based on worst-case rounding errors. I use this as my tolerance for collapsing edges, since I can't trust geometry smaller than this tolerance. Looking at the old version of this function, I have the feeling I had a bug already. Would be great to have another set of eyes on this.

Reply:
`newScale ~ oldScale * transform_`. It seems to me that you are multiplying the factor twice here. Maybe you should multiply `precision` by just `newScale/oldScale` or max length?

Reply:
Perhaps so; I think the transform size was always one before since it was getting reset, so maybe that's how it effectively was.